src/Agent.Service/Windows/AgentService.cs

```diff
@@ -145,11 +153,27 @@ protected override void OnStop()
                     if (AgentListener != null && !AgentListener.HasExited)
                     {
                         // Try to let the agent process know that we are stopping
-                        // TODO: This is not working, fix it
-                        AgentListener.StandardInput.WriteLine("\x03");
+                        //Attach service process to console of Agent.Listener process. This is needed,
+                        //because windows service doesn't use its own console.
+                        if (AttachConsole((uint)AgentListener.Id))
+                        {
+                            //Prevent main service process from stopping because of Ctrl + C event with SetConsoleCtrlHandler
+                            SetConsoleCtrlHandler(null, true);
+                            try
+                            {
+                                //Generate console event for current console with GenerateConsoleCtrlEvent (processGroupId should be zero)
+                                GenerateConsoleCtrlEvent(CTRL_C_EVENT, 0);
+                                //Wait for the process to finish (give it up to 30 seconds)
+                                AgentListener.WaitForExit(30000);
+                            }
+                            finally
+                            {
+                                //Disconnect from console and restore Ctrl+C handling by main process
+                                FreeConsole();
```

Comment:
I forgot win32 apis don't throw and you need to check error explicitly.

looks good.